pkg/gadgets/traceloop/tracer/perf_buffer.go

```diff
@@ -0,0 +1,250 @@
+//go:build linux
+// +build linux
+
+// Copyright 2019-2022 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tracer
+
+import (
+	"fmt"
+	"reflect"
+	"sync/atomic"
+	"unsafe"
+
+	"github.com/cilium/ebpf/perf"
+	log "github.com/sirupsen/logrus"
+
+	"golang.org/x/sys/unix"
+)
+
+type perfEventHeader struct {
+	Type uint32
+	Misc uint16
+	Size uint16
+}
+
+type perfEventSample struct {
+	Size uint32
+}
+
+func getRingsNumber(perfReader *perf.Reader) int {
+	v := reflect.ValueOf(perfReader)
+	rings := reflect.Indirect(v).FieldByName("rings")
+
+	return rings.Len()
+}
+
+func readOverWritable(reader *reader, callback func(record perf.Record, size uint32) error) error {
+	// Get private field with reflect.
+	v := reflect.ValueOf(reader.perfReader)
+	rings := reflect.Indirect(v).FieldByName("rings")
+	ringsNumber := rings.Len()
+
+	// For each perf buffer, we have as many buffers than CPU.
+	for i := 0; i < ringsNumber; i++ {
+		rr := reflect.Indirect(rings.Index(i)).FieldByName("ringReader")
```

Comment:
This was crashing on my system. After some investigation I found that some of the rings can be nil if the CPU is offline -> https://github.com/cilium/ebpf/blob/a9cf21df64c2962c130e2f1b12d467d2fba0996a/perf/reader.go#L216. 

Something like fixes the problem:
```diff
diff --git pkg/gadgets/traceloop/tracer/perf_buffer.go pkg/gadgets/traceloop/tracer/perf_buffer.go
index b5103fa5..c4e74738 100644
--- pkg/gadgets/traceloop/tracer/perf_buffer.go
+++ pkg/gadgets/traceloop/tracer/perf_buffer.go
@@ -54,8 +54,12 @@ func readOverWritable(reader *reader, callback func(record perf.Record, size uin
 
        // For each perf buffer, we have as many buffers than CPU.
        for i := 0; i < ringsNumber; i++ {
-               rr := reflect.Indirect(rings.Index(i)).FieldByName("ringReader")
-
+               ring := rings.Index(i)
+               if ring.IsNil() {
+                       // The CPU is offline, skip it.
+                       continue
+               }
+               rr := reflect.Indirect(ring).FieldByName("ringReader")
                m := reflect.Indirect(rr).FieldByName("meta")
                meta := (*unix.PerfEventMmapPage)(unsafe.Pointer(m.Pointer()))

```

Reply:
I have no idea:

```
$ cat /sys/devices/system/cpu/possible 
0-31
$ cat /sys/devices/system/cpu/online
0-15
```